File: app/src/main/java/com/bitflaker/lucidsourcekit/database/dreamjournal/daos/JournalEntryHasTagDao.java
package com.bitflaker.lucidsourcekit.database.dreamjournal.daos;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;

import com.bitflaker.lucidsourcekit.database.dreamjournal.entities.JournalEntryHasTag;
import com.bitflaker.lucidsourcekit.database.dreamjournal.entities.resulttables.TagCount;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;

@Dao
public interface JournalEntryHasTagDao {
    @Query("SELECT JournalEntryTag.description as tag, COUNT(JournalEntryHasTag.tagId) as count FROM JournalEntryHasTag LEFT JOIN JournalEntryTag ON JournalEntryHasTag.tagId = JournalEntryTag.tagId LEFT JOIN JournalEntry ON JournalEntryHasTag.entryId = JournalEntry.entryId WHERE JournalEntry.timeStamp BETWEEN :startTimestamp AND :endTimestamp GROUP BY JournalEntryHasTag.tagId ORDER BY count DESC LIMIT :limit")
    Single<List<TagCount>> getMostUsedTagsList(long startTimestamp, long endTimestamp, int limit);

    @Insert
    Single<List<Long>> insertAll(List<JournalEntryHasTag> journalEntryHasTags);

    @Delete
    Completable delete(JournalEntryHasTag journalEntryHasTag);

    @Query("DELETE FROM JournalEntryHasTag WHERE entryId = :entryId")
    Completable deleteAll(int entryId);
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/database/dreamjournal/daos/JournalEntryHasTagDao.kt
package com.bitflaker.lucidsourcekit.database.dreamjournal.daos

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import com.bitflaker.lucidsourcekit.database.dreamjournal.entities.JournalEntryHasTag
import com.bitflaker.lucidsourcekit.database.dreamjournal.entities.resulttables.TagCount

@Dao
interface JournalEntryHasTagDao {
    @Query("SELECT JournalEntryTag.description as tag, COUNT(JournalEntryHasTag.tagId) as count FROM JournalEntryHasTag LEFT JOIN JournalEntryTag ON JournalEntryHasTag.tagId = JournalEntryTag.tagId LEFT JOIN JournalEntry ON JournalEntryHasTag.entryId = JournalEntry.entryId WHERE JournalEntry.timeStamp BETWEEN :startTimestamp AND :endTimestamp GROUP BY JournalEntryHasTag.tagId ORDER BY count DESC LIMIT :limit")
    suspend fun getMostUsedTagsList(startTimestamp: Long, endTimestamp: Long, limit: Int): List<TagCount>

    @Insert
    suspend fun insertAll(journalEntryHasTags: List<JournalEntryHasTag>): List<Long>

    @Delete
    suspend fun delete(journalEntryHasTag: JournalEntryHasTag)

    @Query("DELETE FROM JournalEntryHasTag WHERE entryId = :entryId")
    suspend fun deleteAll(entryId: Int)
}
